kong/tools/wrpc/threads.lua

```diff
@@ -0,0 +1,99 @@
+local pb = require "pb"
+local utils = require "kong.tools.wrpc.utils"
+local message = require "kong.tools.wrpc.message"
+
+local safe_args = utils.safe_args
+local endswith = utils.endswith
+local pb_decode = pb.decode
+
+local ngx_now = ngx.now
+local ngx_log = ngx.log
+local NOTICE = ngx.NOTICE
+local exiting = ngx.worker.exiting
+local sleep = ngx.sleep
+local thread_spawn = ngx.thread.spawn
+local thread_kill = ngx.thread.kill
+local thread_wait = ngx.thread.wait
+
+local process_message = message.process_message
+local handle_error = message.handle_error
+
+local _M = {}
+
+local function step(wrpc_peer)
+  local msg, err = wrpc_peer:receive()
+
+  while msg ~= nil do
+    msg = assert(pb_decode("wrpc.WebsocketPayload", msg))
+    assert(msg.version == "PAYLOAD_VERSION_V1", "unknown encoding version")
+    local payload = msg.payload
+
+    if payload.mtype == "MESSAGE_TYPE_ERROR" then
+      handle_error(wrpc_peer, payload)
+
+    elseif payload.mtype == "MESSAGE_TYPE_RPC" then
+      local ack = payload.ack or 0
+      local deadline = payload.deadline or 0
+
+      if ack == 0 and deadline < ngx_now() then
+        ngx_log(NOTICE, "[wRPC] Expired message (", deadline, "<", ngx_now(), ") discarded")
+
+      elseif ack ~= 0 and deadline ~= 0 then
+        ngx_log(NOTICE, "[WRPC] Invalid deadline (", deadline, ") for response")
+
+      else
+        process_message(wrpc_peer, payload)
+      end
+    end
+
+    msg, err = wrpc_peer:receive()
+  end
+
+  if err ~= nil and not endswith(err, "timeout") then
+    ngx_log(NOTICE, "[wRPC] WebSocket frame: ", err)
+    wrpc_peer.closing = true
+  end
+end
+
+function _M.spawn(wrpc_peer)
+  wrpc_peer._receiving_thread = assert(thread_spawn(function()
+    while not exiting() and not wrpc_peer.closing do
+      step(wrpc_peer)
+      sleep(0)
+    end
```

Comment:
This looks dangerously close to an active waiting. What would be the downside of sleeping for longer instead of 0 (perhaps some configurable value)?

Reply:
`step()` does block on I/O, so the sleep might not be that important.  perhaps only do it if there was an error?  maybe even move it (the sleep) inside `step()`?

Reply:
I will do this:

```suggestion
    while not exiting() and not wrpc_peer.closing do
      local ok = step(wrpc_peer)
      -- something wrong with this step
      -- let yield instead of retry immediately
      if not ok then
        sleep(0)
      end
    end
```